### PR TITLE
Fix flush_cache issues in yum_package

### LINKED ISSUE #4691: slow performance in yum_package provider in 12.7.x through 12.9.x
## Description

* running chef-solo from the chef11 gem completes in roughly 14 seconds
* running chef-solo from the chef12 gem completes in over 2 minutes

## Chef Version

11.8.2 -> 12.8.1

## Platform Version

CentOS release 6.7 (Final)

## Replication Case

run the exact same recipes on the exact same hardware under both versions. 

behind the scenes, i noticed that yum-dump.py was running a LOT more under chef12. it appears to run on every single package idempotency check.

## Client Output

https://gist.github.com/downneck/d019bfa4e9cccc65f7ad

## Stacktrace

Please include the stacktrace.out output or link to a gist of it, if there is one.

### PR DESCRIPTION
**Pull request by jaymzh**
Fixes #4691. I hope.

### CODE REVIEW

File: lib/chef/resource/yum_package.rb [hunk 1]
```diff
@@ -30,13 +30,13 @@ class YumPackage < Chef::Resource::Package
       property :flush_cache, Hash, default: { before: false, after: false }, coerce: proc { |v|
         # TODO these append rather than set. This is probably wrong behavior, but we're preserving it until we know
         if v.is_a?(Array)
-          v.each { |arg| flush_cache[arg] = true }
-          flush_cache
+          x = {}
+          v.each { |arg| x[arg] = true }
```
Comment (lamont-granquist) [inline lib/chef/resource/yum_package.rb]: should probably be each_with_object or collect

Comment (lamont-granquist) [inline lib/chef/resource/yum_package.rb]: also you're breaking the behavior in the TODO (which i think is fine)

File: lib/chef/resource/yum_package.rb [hunk 2]
```diff
@@ -30,13 +30,13 @@ class YumPackage < Chef::Resource::Package
...
+          x
         elsif v.any?
           v
         else
-          # TODO calling flush_cache({}) does a get instead of a set. This is probably wrong behavior, but we're preserving it until we know
-          flush_cache
+          { before: v, after: v}
```
Comment (lamont-granquist) [inline lib/chef/resource/yum_package.rb]: shouldn't this just be v?  although i think it makes my brain hurt...   i don't think anything you do here turns this from a getter into a setter does it?

i suspect nobody uses flush_cache({}) as a no-op getter.

this code is now generally trying way to hard to be backcompat and screwing up everything in the process, i'd sort of scrap it and rewrite it cleanly, then ship fixes if we actually do break anyone (because currently we've broken /everyone/ and the code is very hard to follow).

**Comment by lamont-granquist** [comment]
yeah i think delete the two TODO lines and delete the else clause entirely.

my bet is that doing so doesn't cause any regression.

if it does cause a regression, it'll be way less bad than the regression we're currently suffering under.

if we get that regression, then we can pick the best way to fix it after we see it.

**Comment by jaymzh** [comment]
I don't think we can delete the else clause as we have to make it a hash if if it's not. But I'll nuke the todo.

File: lib/chef/resource/yum_package.rb (hunk 1 above)
Comment (jaymzh) [inline lib/chef/resource/yum_package.rb]: collect won't work cause need a hash, but each_with_object is fine (and the same)

**Comment by jaymzh** [comment]
When I change this code it fixes it, but monkey patching it from a library doesn't seem to... not sure why. Still playing.

**Comment by jaymzh** [comment]
OK @lamont-granquist I don't think I can monkey patch that using `property` keyword... property does magic...  so either I need to get a version cut with this ASAP or I need help with magic above my head :)

**Comment by lamont-granquist** [comment]
uh hrm, yeah i suspect property does do lots of magic.

dunno what to tell you about getting new versions cut, i think we're still about 2 weeks-ish away from the next scheduled release based on when the last one went out.  beat on your support contract a bit and you might make it happen tho.  although we've got a bad outstanding regression in master i just cut today #5257, but i don't think it'll be anything insurmountable to fix, just another bug.

**Comment by jaymzh** [comment]
`chef_gem` is evil, we don't use it so not worried. Installing gems via `gem` on prod systems is asinine.

Any ideas on how to monkey patch wahtever methods are being jammed into the `YumPackage` class? A new release seems unlikely but we're a bit fucked here. ANd I CC'd you on the ZenDesk ticket already.

**Comment by lamont-granquist** [comment]
so we don't document using this property as anything other than a hash or an array.  i don't think boolean true/false have been an option?

although i don't mind adding that -- but i thought that else clause was intended to only catch the empty-hash '{}' case and attempt to coerce it into returning the same value...

**Comment by lamont-granquist** [comment]
and no, my suggestion would just be to open the class and copypasta this one 'line' into it...

its possible coerce arguments get pushed into an array and you need to blow away the array?

**Comment by lamont-granquist** [comment]
`Chef::Resource::YumPackage.properties.delete(:flush_cache)`?

**Comment by jaymzh** [comment]
Haha I tried something simialr, but it doesn't work yet:

```
  Chef::Resource::YumPackage.properties[:flush_cache][:coerce] = proc { |v|
          if v.is_a?(Array)
            v.each_with_object({}) { |arg, obj| obj[arg] = true }
          elsif v.any?
            v
          else
            { before: v, after: v}
          end
          }
```

**Comment by lamont-granquist** [comment]
yeah i'd try just nuking the property and then using the DSL to rebuild it all...

**Comment by jaymzh** [comment]
Ha! This does it:

```
  opts = Chef::Resource::YumPackage.properties[:flush_cache].options
  opts[:coerce] = proc { |v|
          if v.is_a?(Array)
            v.each_with_object({}) { |arg, obj| obj[arg] = true }
          elsif v.any?
            v
          else
            { before: v, after: v}
          end
          }
  Chef::Resource::YumPackage.properties[:flush_cache] = opts
```

**Comment by jaymzh** [comment]
delete doesn't work:

```
chef (12.13.37)> Chef::Resource::YumPackage.properties.delete(:flush_cache)
 => #<Chef::Property:0x00000000e03b48 @options={:default=>{:before=>false, :after=>false}, :coerce=>#<Proc:0x00000000e04598@/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.13.37/lib/chef/resource/yum_package.rb:30>, :instance_variable_name=>:@flush_cache, :name=>:flush_cache, :declared_in=>Chef::Resource::YumPackage, :is=>Hash}> 

chef (12.13.37)> Chef::Resource::YumPackage.properties[:flush_cache]
 => #<Chef::Property:0x00000000e03b48 @options={:default=>{:before=>false, :after=>false}, :coerce=>#<Proc:0x00000000e04598@/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.13.37/lib/chef/resource/yum_package.rb:30>, :instance_variable_name=>:@flush_cache, :name=>:flush_cache, :declared_in=>Chef::Resource::YumPackage, :is=>Hash}> 
```

**Comment by jaymzh** [comment]
OK this seems to work....
```
  Chef::Log.info('Monkey-patching yum_package resource')
  opts = Chef::Resource::YumPackage.properties[:flush_cache].options
  opts[:coerce] = proc { |v|
          if v.is_a?(Array)
            v.each_with_object({}) { |arg, obj| obj[arg] = true }
          elsif v.any?
            v
          else
            { before: v, after: v}
          end
          }
  Chef::Resource::YumPackage.property(:flush_cache, opts)
```

If I specify a type, the validator becomes `... is => [Hash, Hash]` which is weird.

**Comment by jaymzh** [comment]
OK, @lamont-granquist pushing that out to the 12.13.37 hosts. We'll see how that goes. You OK merging this meantime?

**Comment by jaymzh** [comment]
replaced {} with do/end cause... style.

**Comment by jaymzh** [comment]
I just figured out the actual problem!!!!

On the tier effected one of the earliest packages was setting 'flush_cache [:before]` - and since the bug was that it was additive and not per-package... that's what fucked us.

**Comment by jaymzh** [comment]
wtf is this:

```
/home/travis/build/chef/chef/lib/chef/resource/yum_package.rb:31:in `proc': tried to create Proc object without a block (ArgumentError)
```

that's not true.

**Comment by lamont-granquist** [comment]
whoah it wasn't per-package?  so it like updated the default value more or less?  yikes...

**Comment by lamont-granquist** [comment]
yeah and i was just gonna suggest do..end, its different in subtle and annoying ways...

**Comment by jaymzh** [comment]
so `do..end` breaks things. I'll go back to `{}` because... I don't even know. rubocop requires do..end for multi-line blocks. and do..end *works* but it chokes rspec. so whatever.

**Comment by jaymzh** [comment]
fucking hell ruby

**Comment by jaymzh** [comment]
and yes, it was effectively updating the default

**Comment by jaymzh** [comment]
from 2550 second runs to 120 second runs. That's better. :)

**Comment by lamont-granquist** [comment]
that explains why i couldn't replicate either, nobody reported that you had to call flush_cache once and then it flushed the cache every time after that...

**Comment by jaymzh** [comment]
wasn't exactly easy to figure out :) @joshuamiller01 mentioned it was only happening on one of our tiers, so I grepped the code-base for `flush_cache` and then saw it was only in that cookbook, and that combined with the "we're appending, which seems odd" TODO comment, made it all click.

**Comment by jaymzh** [comment]
OK, @lamont-granquist all the tests pass. I'm gonna have a fucking drink now.

**Comment by lamont-granquist** [comment]
🍺

**Comment by jaymzh** [comment]
CC @chef/client-core just for coverage.

**Comment by lamont-granquist** [comment]
i'm 👍 

since its friday late, lets hang off on merging till everyone else can look at it

**Comment by lamont-granquist** [comment]
(and for the record this is way too tortured to write a spec test for, i doubt it'll ever recur)

**Comment by coderanger** [comment]
👍 Accidentally mutable defaults bite us again :-( Maybe for bonus "and the horse you rode in on" call `.freeze` on the default value?

**Comment by thommay** [comment]
oof :( thanks for the debugging and nailing this, @jaymzh  👍
